sshd-core/src/main/java/org/apache/sshd/common/session/helpers/KeyExchangeMessageHandler.java

```diff
@@ -0,0 +1,485 @@
+import java.util.List;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Supplier;
+
+import org.apache.sshd.common.SshConstants;
+import org.apache.sshd.common.future.DefaultKeyExchangeFuture;
+import org.apache.sshd.common.io.IoWriteFuture;
+import org.apache.sshd.common.kex.KexState;
+import org.apache.sshd.common.util.ExceptionUtils;
+import org.apache.sshd.common.util.ValidateUtils;
+import org.apache.sshd.common.util.buffer.Buffer;
+import org.apache.sshd.common.util.threads.ThreadUtils;
+import org.slf4j.Logger;
+
+/**
+ * Manages SSH message sending during a key exchange. RFC 4253 specifies that during a key exchange, no high-level
+ * messages are to be sent, but a receiver must be able to deal with messages "in flight" until the peer's
+ * {@link SshConstants#SSH_MSG_KEX_INIT} message is received.
+ * <p>
+ * Apache MINA sshd queues up high-level messages that threads try to send while a key exchange is ongoing, and sends
+ * them once the key exchange is done. Sending queued messages may make the peer re-trigger a new key exchange, in which
+ * case sending queued messages stops and is resumed at the end of the new key exchange.
+ * </p>
+ *
+ * @see <a href="https://tools.ietf.org/html/rfc4253#section-7">RFC 4253</a>
+ */
+public class KeyExchangeMessageHandler {
+
+    // With asynchronous flushing we get a classic producer-consumer problem. The flushing thread is the single
+    // consumer, and there is a risk that it might get overrun by the producers. The classical solution of using a
+    // LinkedBlockingQueue with a fixed maximum capacity doesn't work: we cannot make the producers block when the queue
+    // is full; we might deadlock or be unable to handle any incoming message.
+    //
+    // We need an unbounded queue that never blocks the producers, but that manages to throttle them such that the
+    // flushing thread can actually finish, and we still can handle incoming messages (in particular also the peer's
+    // SSH_MSG_NEW_KEYS, since we start flushing already after having sent our own SSH_MSG_NEW_KEYS).
+    //
+    // This is achieved by giving the flushing thread priority over the threads that might enqueue additional packets
+    // and flushing at least two packets at a time. Additionally the flush loop releases and shortly afterwards
+    // re-acquires the write lock, so normally not many readers (i.e., writePacket() calls) will get a chance to enqueue
+    // new packets.
+
+    /**
+     * We need the flushing thread to have priority over writing threads. So we use a lock that favors writers over
+     * readers, and any state updates and the flushing thread are writers, while writePacket() is a reader.
+     */
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock(false);
```

Comment:
Let's make the members *protected* in the spirit of letting our users override whatever they like - e,g, provide their own derived handler.

Reply:
Done.